frontend/model/contracts/mailbox.js

```diff
@@ -98,63 +127,108 @@ sbp('chelonia/defineContract', {
       }
     },
     'gi.contracts/mailbox/leaveDirectMessage': {
-      validate: (data, { state, meta }) => {
-        objectOf({
-          username: string
-        })(data)
+      validate: (data, { state, getters, meta }) => {
+        objectOf({ username: string })(data)
         if (state.attributes.creator !== meta.username) {
           throw new TypeError(L('Only the mailbox creator can leave direct message channel.'))
-        } else if (!state.users[data.username]?.joinedDate) {
+        } else if (!getters.ourOneToOneDirectMessages[data.username]?.joinedDate) {
           throw new TypeError(L('Not joined or already left direct message channel.'))
         }
       },
-      process ({ data }, { state }) {
-        Vue.set(state.users[data.username], 'joinedDate', null)
+      process ({ data }, { state, getters }) {
+        const contractID = getters.ourOneToOneDirectMessages[data.username].contractID
+        Vue.set(state.chatRooms[contractID], 'joinedDate', null)
       },
-      sideEffect ({ contractID, data }, { state }) {
-        leaveChatRoom({ contractID: state.users[data.username].contractID })
+      sideEffect ({ data }, { getters }) {
+        leaveChatRoom({ contractID: getters.ourOneToOneDirectMessages[data.username].contractID })
+      }
+    },
+    'gi.contracts/mailbox/createGroupMessage': {
+      validate: (data, { state, meta }) => {
+        objectOf({ contractID: string })(data)
+        if (state.attributes.creator !== meta.username) {
+          throw new TypeError(L('Only the mailbox creator can create group message channel.'))
+        } else if (state.chatRooms[data.contractID]) {
+          throw new TypeError(L('Already existing group message channel.'))
+        }
+      },
+      process ({ meta, data }, { state }) {
+        Vue.set(state.chatRooms, data.contractID, {
+          hidden: false,
+          joinedDate: meta.createdDate
+        })
+      },
+      async sideEffect ({ contractID, data }) {
+        await sbp('chelonia/contract/sync', data.contractID)
+
+        if (!sbp('chelonia/contract/isSyncing', contractID)) {
+          await sbp('controller/router')
+            .push({ name: 'GroupChatConversation', params: { chatRoomId: data.contractID } })
+            .catch(logExceptNavigationDuplicated)
+        }
+      }
+    },
+    'gi.contracts/mailbox/joinGroupMessage': {
+      validate: (data, { state, getters, meta }) => {
+        objectOf({ contractID: string })
+        // TODO: need to validate if meta.username is part of the chatroom of contractID
+      },
+      process ({ meta, data }, { state }) {
+        if (state.chatRooms[data.contractID]) {
+          throw new TypeError(L('Already existing group message channel.'))
+        }
+
+        const joinedDate = state.attributes.autoJoinAllowance ? meta.createdDate : null
+        Vue.set(state.chatRooms, data.contractID, {
+          hidden: false, // TODO: this hidden attribute should be there in state.js as global
```

Comment:
What does this comment mean?

Remember - global state in state.js cannot be used here, since inside of contracts the only state that can be used is the state that's passed in via the parameters to `process` (and constants that are defined in the contract file itself).

Reply:
Yeah. Correct, we can not use the global state inside the contract state because it could be different from user to user.
What I meant is that we should define `hidden` value in the global state not in the contract state.
I will re-write the comment or remove the `hidden` attribute then. Because `hidden` attribute is not using at the moment.

Reply:
I thought wrong. This shouldn't be there in global state. I will re-write the comment.

Reply:
If you mean that the default is defined in state.js, and then it is passed in via the parameters to `process` (because it's included as part of the `data` in the message, then that makes sense 👍

Reply:
Let's think about the workflow once more.

If I hide the direct message by clicking 'X' button or any other, that direct message will be invisible in the current device.
Then, I login in another device and check the direct message visibility.
What should it be? `Visible` or `Invisible`. I think it should be `Invisible`. So the `hidden` attribute should be there inside the `mailbox` contract I think.

That's why I think I should re-write the comment.

```
hidden: false, // NOTE: this attr is used to hide/show direct messages
```

Reply:
Yes, that sounds good 👍